controllers/automation/packagedeployment_controller.go

```diff
@@ -434,3 +530,171 @@ func (r *PackageDeploymentReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	upstreamLock := sourcePR.Status.UpstreamLock
+	repoList := configapi.RepositoryList{}
+	err := r.PorchClient.List(ctx, &repoList, &client.ListOptions{})
+	if err != nil {
+		return nil
+	}
+	return r.availableUpdates(ctx, upstreamLock, &repoList)
+}
+
+func (r *PackageDeploymentReconciler) updateDraft(ctx context.Context,
+	sourcePR *porchv1alpha1.PackageRevision,
+	newUpstreamPR *porchv1alpha1.PackageRevision) (*porchv1alpha1.PackageRevision, error) {
+
+	oldPR := sourcePR.DeepCopy()
+	tasks := sourcePR.Spec.Tasks
+	cloneTask := tasks[0]
+	updateTask := porchv1alpha1.Task{
+		Type: porchv1alpha1.TaskTypeUpdate,
+		Update: &porchv1alpha1.PackageUpdateTaskSpec{
+			Upstream: cloneTask.Clone.Upstream,
+		},
+	}
+	updateTask.Update.Upstream.UpstreamRef.Name = newUpstreamPR.Name
+	oldPR.Spec.Tasks = append(oldPR.Spec.Tasks, updateTask)
+
+	err := r.PorchClient.Update(ctx, oldPR)
+	if err != nil {
+		return nil, err
+	}
+	return oldPR, nil
+}
+
+// Updating a published package revision means making a copy of it and updating the
+// newly created draft.
+func (r *PackageDeploymentReconciler) updateLatest(ctx context.Context,
+	sourcePR *porchv1alpha1.PackageRevision) (*porchv1alpha1.PackageRevision, error) {
+	upstreamPR := r.discoverUpdate(ctx, sourcePR)
+	if upstreamPR == nil {
+		return sourcePR, nil
+	}
+	newPR := &porchv1alpha1.PackageRevision{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "PackageRevision",
+			APIVersion: porchv1alpha1.SchemeGroupVersion.Identifier(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: sourcePR.Namespace,
+		},
+		Spec: sourcePR.Spec,
+	}
+
+	newPR.Spec.Revision = ""
+	wsNumStr := strings.TrimPrefix(string(newPR.Spec.WorkspaceName), "packagedeployment-")
+	wsNum, _ := strconv.Atoi(wsNumStr)
+	wsNum++
+	newPR.Spec.WorkspaceName = porchv1alpha1.WorkspaceName(fmt.Sprintf("packagedeployment-%d", wsNum))
+	newPR.Spec.Lifecycle = porchv1alpha1.PackageRevisionLifecycleDraft
+
+	if err := r.PorchClient.Create(ctx, newPR); err != nil {
+		return nil, err
+	}
+
+	if err := r.loadPackageRevisions(ctx); err != nil {
+		return nil, err
+	}
+
+	// get the newly created package back
+	if byRepo, ok := r.packageRevs[newPR.ObjectMeta.Namespace]; ok {
+		if byPkg, ok := byRepo[newPR.Spec.RepositoryName]; ok {
+			if byRev, ok := byPkg[newPR.Spec.PackageName]; ok {
+				drafts := byRev[""]
+				for _, draft := range drafts {
+					if draft.Spec.WorkspaceName == newPR.Spec.WorkspaceName {
+						return r.updateDraft(ctx, draft, upstreamPR)
+					}
+				}
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("could not find newly created package revision")
+}
+
+func (r *PackageDeploymentReconciler) availableUpdates(ctx context.Context, upstreamLock *porchv1alpha1.UpstreamLock, repositories *configapi.RepositoryList) *porchv1alpha1.PackageRevision {
+	// This code is largely copied from the kpt repo.
+	// We can probably refactor the kpt code into a library so that clients can consume it directly.
```

Comment:
I'm not sure what the right protocol is here, but I can move this into some `third_party` folder if necessary, since large parts of this were copied.

Reply:
I think it's ok, probably need to reference the actual file or at least the repo. It's all under Apache license anyway I believe, so sharing with attribution is fine.

Reply:
That makes sense, thanks! I added the link to the copied code in the comment.